ERCS/erc-7943.md

```diff
@@ -0,0 +1,518 @@
+- Burning CAN run validation to prohibit burning more assets than the unfrozen amount. For example, in public burning functions.
+- At the same time, burning CAN bypass freezing validation, in which case it MUST update the freezing status if required, emitting a `Frozen` event before the `Transfer` one. For example, in permissioned burning functions.
+
+The `ERC7943NotAllowedTransfer` and `ERC7943NotAllowedUser` errors CAN be used as a general revert mechanism whenever internal calls to `isUserAllowed` or `isTransferAllowed` return false. Those MIGHT NOT be used or MIGHT be replaced by more specific errors depending on the custom checks performed inside those calls.
+
+In general, error specificity is prioritized, meaning that broad errors such as `ERC7943NotAllowedTransfer` SHOULD be thrown if more specific ones such as `ERC7943InsufficientUnfrozenBalance` do not apply. The same is true for `ERC7943InsufficientUnfrozenBalance` that SHOULD be triggered when a transfer is attempted from `user` with an `amount` of `tokenId` less or equal than its balance, but greater than its unfrozen balance. But if the amount is greater than the whole balance, unrelated from frozen amount, more specific errors from the base standard SHOULD be used instead, like `ERC20InsufficientBalance`.
+
+## Rationale
+
+*   **Minimalism:** Defines only the essential functions (`forceTransfer`, `setFrozen`, `isUserAllowed`, `isTransferAllowed`, `getFrozen`) and associated events/errors needed for common RWA compliance and control patterns, avoiding mandated complexity or opinionated features.
+*   **Flexibility:** Provides standard view functions (`isUserAllowed`, `isTransferAllowed`, `getFrozen`) for compliance checks without dictating *how* those checks are implemented internally by the token contract. This allows diverse compliance strategies.
+*   **Compatibility:** Designed as an interface layer compatible with existing base standards like [ERC-20](./eip-20.md), [ERC-721](./eip-721.md) and [ERC-1155](./eip-1155.md). Implementations extend from [ERC-7943](./eip-7943) alongside their base standard interface.
+*   **RWA Essential:** Includes `forceTransfer` and `setFrozen` as standard functions, acknowledging its importance for regulatory enforcement in the RWA space, distinct from standard transfers. Mandates access control for this sensitive function.
+*   **[ERC-165](./eip-165.md):** Ensures implementing contracts can signal support for this interface.
+
+As an example, a Uniswap v4 pool can integrate with [ERC-7943](./eip-7943.md) based [ERC-20](./eip-20.md) tokens by calling `isUserAllowed` or `isTransferAllowed` within its before/after hooks to handle these assets in a compliant manner. Users can then expand these tokens with additional features to fit the specific needs of individual asset types, either with on-chain identity systems, historical balances tracking for dividend distributions, semi-fungibility with token metadata, and other custom functionalities.
```

Comment:
Try to avoid mentioning specific products. It gives products that exist today an unfair advantage over products that come later (because ERCs eventually become immutable, so newcomers cannot be added).

Reply:
Thanks, rephrased !